File: server/internal/sync/merge.go
package sync

import "encoding/json"

// FieldClock is a row's field clock: field name -> HLC of the last write.
type FieldClock map[string]string

func ParseFieldClock(s string) FieldClock {
	fc := FieldClock{}
	if s != "" {
		_ = json.Unmarshal([]byte(s), &fc)
	}
	return fc
}

func (fc FieldClock) Marshal() string {
	b, _ := json.Marshal(fc)
	return string(b)
}

// Accept decides per field: if the incoming HLC wins, the field is
// applied and the field clock updated.
func (fc FieldClock) Accept(field, incomingHLC string) bool {
	if Compare(incomingHLC, fc[field]) > 0 {
		fc[field] = incomingHLC
		return true
	}
	return false
}

// --- OR-Set (Observed-Remove Set, add-wins) ---
//
// JSON representation: { "<element>": { "a": ["<tag>",...], "r": ["<tag>",...] } }
// An element is visible while it has an add tag that has not been removed
// (observed) wurde.

type ORSet map[string]struct {
	A []string `json:"a"`
	R []string `json:"r"`
}

func ParseORSet(s string) ORSet {
	set := ORSet{}
	if s != "" && s != "null" {
		_ = json.Unmarshal([]byte(s), &set)
	}
	return set
}

func (s ORSet) Marshal() string {
	b, _ := json.Marshal(s)
	return string(b)
}

// Add inserts an element with a unique tag (HLC).
func (s ORSet) Add(elem, tag string) {
	e := s[elem]
	e.A = appendUnique(e.A, tag)
	s[elem] = e
}

// Remove only deletes the currently observed add tags (add-wins against
// concurrent, not-yet-observed adds).
func (s ORSet) Remove(elem string) {
	e := s[elem]
	for _, t := range e.A {
		e.R = appendUnique(e.R, t)
	}
	s[elem] = e
}

// Merge vereinigt zwei OR-Sets (kommutativ, idempotent).
func (s ORSet) Merge(other ORSet) {
	for elem, oe := range other {
		e := s[elem]
		for _, t := range oe.A {
			e.A = appendUnique(e.A, t)
		}
		for _, t := range oe.R {
			e.R = appendUnique(e.R, t)
		}
		s[elem] = e
	}
}

// Values returns the currently visible elements.
func (s ORSet) Values() []string {
	var out []string
	for elem, e := range s {
		rm := map[string]bool{}
		for _, t := range e.R {
			rm[t] = true
		}
		for _, t := range e.A {
			if !rm[t] {
				out = append(out, elem)
				break
			}
		}
	}
	return out
}

func appendUnique(xs []string, x string) []string {
	for _, v := range xs {
		if v == x {
			return xs
		}
	}
	return append(xs, x)
}

File: app/src/lib/local/merge.ts
// Mirror of the server merge logic (internal/sync/merge.go) for the client.
import { compare } from './hlc';

// --- Per-field-LWW ---
export type FieldClock = Record<string, string>;

export function parseFieldClock(s: string | null | undefined): FieldClock {
  if (!s) return {};
  try { return JSON.parse(s); } catch { return {}; }
}

// Applies the field only if the incoming HLC is newer.
export function accept(fc: FieldClock, field: string, hlc: string): boolean {
  if (compare(hlc, fc[field] ?? '') > 0) {
    fc[field] = hlc;
    return true;
  }
  return false;
}

// --- OR-Set (add-wins) ---
type Elem = { a: string[]; r: string[] };
export type ORSet = Record<string, Elem>;

export function parseORSet(s: string | null | undefined): ORSet {
  if (!s || s === 'null') return {};
  try { return JSON.parse(s); } catch { return {}; }
}

const uniq = (xs: string[], x: string) => (xs.includes(x) ? xs : [...xs, x]);

export function orAdd(set: ORSet, elem: string, tag: string) {
  const e = set[elem] ?? { a: [], r: [] };
  e.a = uniq(e.a, tag);
  set[elem] = e;
}
export function orRemove(set: ORSet, elem: string) {
  const e = set[elem] ?? { a: [], r: [] };
  for (const t of e.a) e.r = uniq(e.r, t); // nur observed tags remove
  set[elem] = e;
}
export function orMerge(into: ORSet, other: ORSet) {
  for (const [elem, oe] of Object.entries(other)) {
    const e = into[elem] ?? { a: [], r: [] };
    for (const t of oe.a) e.a = uniq(e.a, t);
    for (const t of oe.r) e.r = uniq(e.r, t);
    into[elem] = e;
  }
}
export function orValues(set: ORSet): string[] {
  const out: string[] = [];
  for (const [elem, e] of Object.entries(set)) {
    const rm = new Set(e.r);
    if (e.a.some((t) => !rm.has(t))) out.push(elem);
  }
  return out;
}
